images/macos/provision/configuration/CreateCleanAnkaTemplate.ps1

```diff
@@ -0,0 +1,232 @@
+
+    if (-not ($MacOSVersion -in $availableVersions.OsVersion))
+    {
+        Write-Host "Requested macOS version '$MacOSVersion' not found in the list of avaialable installers. Avaialable versions are:`n$($availableVersions.OsVersion)"
+        Write-Host 'Make sure to pass "-BetaSearch $true" if you need a beta version installer'
+        exit 1
+    }
+
+    # Validate there is no softwareupdate at the moment
+    Test-SoftwareUpdate
+
+    $macOsName = ($availableVersions | Where-Object { $_.OsVersion -match $MacOSVersion }).OsName
+    $installerPathPattern = "/Applications/Install*${macOsName}.app"
+
+    if (Test-Path $installerPathPattern) {
+        Write-Host "Removing existed macOS $macOsName installation package before downloading the new one"
+        bash -c "sudo rm -rf $installerPathPattern"
+    }
+
+    Write-Host "Requested macOS version '$MacOSVersion' installer found, fetching it from Apple"
+    softwareupdate --fetch-full-installer --full-installer-version $MacOSVersion
+    $installerPath = Get-ChildItem -Path $installerPathPattern
+    Write-Host "Installer successfully downloaded to '$installerPath'"
+
+    return $installerPath.FullName
+}
+
+function Get-AvailableVersions {
+    param (
+        [Int] $RetryCount = 20,
+        [Int] $RetryInterval = 60,
+        [Bool] $IsBeta = $false
+    )
+
+    if ($IsBeta)
+    {
+        $searchPostfix = " beta"
+    }
+
+    $searchPattern = "macOS (?<osname>[a-zA-z ]*$searchPostfix), Version: (?<osversion>(\d+\.){1,}\d+)"
+
+    while ($RetryCount -ne 0)
+    {
+    $RetryCount--
+    $allVersions = softwareupdate --list-full-installers | Select-String -Pattern $searchPattern -AllMatches | ForEach-Object {
+        $osName = ($_.Matches.Groups | Where-Object { $_.Name -eq "osname" }).Value
+        $osVersion = ($_.Matches.Groups | Where-Object { $_.Name -eq "osversion" }).Value
+        [PSCustomObject]@{
+            OsName = $osName
+            OsVersion = $osVersion
+        }
+    } | Sort-Object -Property OsVersion -Unique
+
+    if ($allVersions)
+    {
+        return $allVersions
+    }
+
+    Write-Host "Could not fetch installers list, wait $RetryInterval seconds, $RetryCount attempts left"
+    Start-Sleep -Seconds $RetryInterval
+    }
+
+    Write-Host "All the attempts exhausted, try again later"
+    exit 1
+}
+
+function Test-SoftwareUpdate {
+    param (
+        [String] $UpdateProcessName = "softwareupdate",
+        [Int] $RetryCount = 20,
+        [Int] $RetryInterval = 60
+    )
+
+    while ($RetryCount -ne 0)
+    {
+        $RetryCount--
+        $updateProcess = (Get-Process -Name $UpdateProcessName -ErrorAction SilentlyContinue).id
+
+        if (-not $updateProcess)
+        {
+            return
+        }
+
+        # Workaround to get commandline param as it doesn't work for macOS atm https://github.com/PowerShell/PowerShell/issues/13943
+        $processName = ps -o command= $updateProcess
```

Comment:
Why do we use multiple approaches to get a process?

Reply:
Do we have any plans to run multiple builds on one node?

Reply:
Potentially yes, there could be two agents on one node so I would prefer to stay on a safe side and check the processes